sdks/java/io/jms/src/main/java/org/apache/beam/sdk/io/jms/JmsIO.java

```diff
@@ -932,36 +1036,158 @@ public void setup() throws Exception {
+          try {
+            // If the connection failed, stopping the connection will throw a JMSException
+            connection.stop();
+          } catch (JMSException exception) {
+            LOG.warn("The connection couldn't be closed", exception);
+          }
+          connection.close();
+          connection = null;
+        }
+      }
+    }
+
+    private static class JmsIOProducerFn<T> extends DoFn<T, T> {
+
+      private final @Initialized JmsConnection<T> jmsConnection;
+      private final TupleTag<T> failedMessagesTag;
+
+      JmsIOProducerFn(JmsIO.Write<T> spec, TupleTag<T> failedMessagesTag) {
+        this.failedMessagesTag = failedMessagesTag;
+        this.jmsConnection = new JmsConnection<>(spec);
+      }
+
+      @StartBundle
+      public void startBundle() throws JMSException {
+        this.jmsConnection.start();
+      }
+
+      @ProcessElement
+      public void processElement(@Element T input, ProcessContext context) {
+        try {
+          this.jmsConnection.publishMessage(input);
+        } catch (JMSException | JmsIOException exception) {
+          context.output(this.failedMessagesTag, input);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle() throws JMSException {
+        this.jmsConnection.close();
+      }
+
+      @Teardown
+      public void tearDown() throws JMSException {
+        this.jmsConnection.close();
+      }
+    }
+
+    static class JmsIOProduceRetryFn<T> extends DoFn<T, T> {
+
+      private transient @Initialized FluentBackoff retryBackOff;
+
+      private final JmsIO.Write<T> spec;
+      private final TupleTag<T> failedMessagesTags;
+      private final @Initialized JmsConnection<T> jmsConnection;
+      private final Counter publicationRetries =
+          Metrics.counter(JMS_IO_PRODUCER_METRIC_NAME, PUBLICATION_RETRIES_METRIC_NAME);
+
+      JmsIOProduceRetryFn(JmsIO.Write<T> spec, TupleTag<T> failedMessagesTags) {
+        this.spec = spec;
+        this.failedMessagesTags = failedMessagesTags;
+        this.jmsConnection = new JmsConnection<>(spec);
+      }
+
+      @Setup
+      public void setup() {
+        RetryConfiguration retryConfiguration = checkStateNotNull(spec.getRetryConfiguration());
+        retryBackOff =
+            FluentBackoff.DEFAULT
+                .withInitialBackoff(checkStateNotNull(retryConfiguration.getInitialDuration()))
+                .withMaxCumulativeBackoff(checkStateNotNull(retryConfiguration.getMaxDuration()))
+                .withMaxRetries(retryConfiguration.getMaxAttempts());
+      }
+
+      @StartBundle
+      public void startBundle() throws JMSException {
+        this.jmsConnection.start();
+      }
+
+      @ProcessElement
+      public void processElement(@Element T input, ProcessContext context) {
+        try {
+          publishMessage(input, context);
+        } catch (IOException | InterruptedException exception) {
+          LOG.error("Error while publishing the message", exception);
```

Comment:
Outputed twice and double logging for failed publishMessage (see below)

publishMessage may only publish messege (no need ProcessContext parameter their) and throw exception when retry count used up. Then deal with the failure in processElement here.

Reply:
Ok, I will do that

Reply:
Something like this

````java
@ProcessElement
public void processElement(@Element T input, ProcessContext context) {
  try {
    publishMessage(input);
  } catch (JMSException | JmsIOException | IOException | InterruptedException exception) {
    LOG.error("Error while publishing the message", exception);
    context.output(this.failedMessagesTags, input);
    if (exception instanceof InterruptedException) {
      Thread.currentThread().interrupt();
    }
  }
}

private void publishMessage(T input)
    throws JMSException, JmsIOException, IOException, InterruptedException {
  Sleeper sleeper = Sleeper.DEFAULT;
  BackOff backoff = checkStateNotNull(retryBackOff).backoff();
  while (true) {
    try {
      this.jmsConnection.publishMessage(input);
      break;
    } catch (JMSException | JmsIOException exception) {
      if (!BackOffUtils.next(sleeper, backoff)) {
        throw exception;
      } else {
        publicationRetries.inc();
      }
    }
  }
}
````

Reply:
yes that looks good